File: sdk/ts/src/integrations/_shared.ts
/**
 * Helpers used by more than one integration. Not part of the public surface.
 */

export function extractMrns(value: unknown): Set<string> {
  const out = new Set<string>();
  walkForMrns(value, out, 0);
  return out;
}

function walkForMrns(value: unknown, out: Set<string>, depth: number): void {
  if (depth > 5) return;
  if (value === null || value === undefined) return;
  if (typeof value === "string") {
    scanStringForMrns(value, out);
    return;
  }
  if (Array.isArray(value)) {
    for (const v of value) walkForMrns(v, out, depth + 1);
    return;
  }
  if (typeof value === "object") {
    const obj = value as Record<string, unknown>;
    const mrn = obj.mrn;
    if (typeof mrn === "string" && mrn) out.add(mrn);
    for (const v of Object.values(obj)) {
      walkForMrns(v, out, depth + 1);
    }
  }
}

// Match any URI-shaped substring. Used to pick MRNs out of free text — for
// example the markdown bodies returned by the Marmot MCP server.
const URI_RE = /\b([a-z][a-z0-9+.-]*):\/\/[^\s'"`)\]\}>,]+/gi;
const NON_MRN_SCHEMES = new Set(["http", "https", "ws", "wss", "ftp", "ftps", "file"]);

function scanStringForMrns(text: string, out: Set<string>): void {
  for (const match of text.matchAll(URI_RE)) {
    const scheme = (match[1] ?? "").toLowerCase();
    if (NON_MRN_SCHEMES.has(scheme)) continue;
    out.add(match[0]);
  }
}

export async function sha256Hex(input: string): Promise<string> {
  const data = new TextEncoder().encode(input);
  const buf = await crypto.subtle.digest("SHA-256", data);
  return Array.from(new Uint8Array(buf), (b) => b.toString(16).padStart(2, "0")).join("");
}

File: sdk/python/src/marmot/integrations/shared.py
"""Helpers used by more than one integration. Not part of the public surface."""

from __future__ import annotations

import hashlib
import json
import re
from typing import Any

# MRN schemes Marmot is known to emit. Conservative — matching ``<scheme>://``
# unconstrained would also catch arbitrary URLs (HTTP, etc.); we only mine for
# schemes we control or that land in the catalog.
MRN_PATTERN = re.compile(
    r"\b(?:mrn|postgres|postgresql|mysql|kafka|s3|gcs|bigquery|snowflake|redis|"
    r"clickhouse|elasticsearch|opensearch|mongodb|dynamodb|airflow|"
    r"dbt|marmot)://[^\s\"'`<>,)\]}]+"
)


def extract_mrns(output: Any) -> set[str]:
    """Walk an arbitrary tool output for asset MRNs.

    Recognises:
    - dicts with an ``mrn`` key,
    - dicts/lists containing such dicts (e.g. ``{"results": [{"mrn": ...}]}``),
    - JSON-encoded strings of any of the above,
    - free text mentioning ``<scheme>://...`` for known catalog schemes
      (markdown bodies returned by the Marmot MCP server fall here),
    - lists of content blocks shaped like ``{"type": "text", "text": ...}``
      (the MCP tool-response envelope).
    """
    found: set[str] = set()
    _walk(output, found, depth=0)
    return found


def _walk(value: Any, out: set[str], *, depth: int) -> None:
    if depth > 5:
        return
    if value is None:
        return
    if isinstance(value, str):
        _walk_string(value, out)
        return
    if isinstance(value, dict):
        mrn = value.get("mrn")
        if isinstance(mrn, str) and mrn:
            out.add(mrn)
        for v in value.values():
            _walk(v, out, depth=depth + 1)
        return
    if isinstance(value, list):
        for v in value:
            _walk(v, out, depth=depth + 1)


def _walk_string(s: str, out: set[str]) -> None:
    """Try to JSON-decode first (most structured tool outputs come back as
    JSON-encoded strings); fall back to regex scanning the raw text."""
    try:
        parsed = json.loads(s)
    except (ValueError, TypeError):
        parsed = None
    if isinstance(parsed, (dict, list)):
        _walk(parsed, out, depth=0)
    for match in MRN_PATTERN.findall(s):
        out.add(match)


def sha256_hex(value: str) -> str:
    return hashlib.sha256(value.encode()).hexdigest()
